csharp/src/Apache.Arrow/C/CArrowSchema.cs

```diff
@@ -0,0 +1,599 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Apache.Arrow.Types;
+
+[UnmanagedFunctionPointer(CallingConvention.StdCall)]
+public delegate void ReleaseCArrowSchema(IntPtr schema);
+
+namespace Apache.Arrow.C
+{
+    /// <summary>
+    /// An Arrow C Data Interface Schema, which represents a type, field, or schema.
+    /// </summary>
+    /// 
+    /// <remarks>
+    /// This is used to export <see cref="ArrowType"/>, <see cref="Field"/>, or
+    /// <see cref="Schema"/> to other languages. It matches the layout of the
+    /// ArrowSchema struct described in https://github.com/apache/arrow/blob/main/cpp/src/arrow/c/abi.h.
+    /// </remarks>
+    [StructLayout(LayoutKind.Sequential)]
+    public struct CArrowSchema
+    {
+        public IntPtr format;
+        public IntPtr name;
+        public IntPtr metadata;
+        public long flags;
+        public long n_children;
+        public IntPtr children;
+        public IntPtr dictionary;
+        [MarshalAs(UnmanagedType.FunctionPtr)]
+        public ReleaseCArrowSchema release;
```

Comment:
We should use function pointers here, not managed delegates.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/unsafe-code#function-pointers

(Coding in GitHub, I'm not 100% sure this code will compile/run, but hopefully you get the idea):

```suggestion
        [MarshalAs(UnmanagedType.FunctionPtr)]
        public delegate* unmanaged<CArrowSchema*, void> release;
```

Reply:
That makes sense. The syntax here is for C# 9.0 and above, so I'll need to figure out the equivalent for C# 8.0.

Reply:
https://github.com/apache/arrow/blob/48df467eb5c684aafbd168ad056c64bb2f6706e8/csharp/Directory.Build.props#L37

just update this to be `latest` instead. Not sure why we are setting `8.0` explicitly.

Reply:
Also, if it makes it easier, we can make the C Data Interface only available on NETCOREAPP. I don't think we need to support it on `netstandard` at all. At least not until someone actually needs it.

Reply:
Can you elaborate on your reasons for targeting `netstandard2.1` instead of `net6.0`? I see that ParquetSharp also targets .NET Framework (`net471`). Would you need the C Data interface on .NET Framework?

Reply:
It's an open source project so we try to support a broad range of platforms, I don't know of specific users that would need .netstandard 2.1 specifically that couldn't use net6.0 though. And no we wouldn't need this supported in net471. So after considering this some more it's probably fine if the C data interface support is netcoreapp only as we could provide arrow integration as a separate library or add a new target and limit support to that